corehq/ex-submodules/couchexport/export.py

```diff
@@ -313,19 +316,28 @@ def get_data(self):
     @classmethod
     def wrap_all_rows(cls, tables):
         """
-        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...))
+        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...)). The rows can be generators.
         """
-        ret = []
-        for name, rows in tables:
-            rows = list(rows)
-            if rows and (not hasattr(rows[0], '__iter__') or isinstance(rows[0], six.string_types)):
-                soft_assert_type_text(rows[0])
+
+        def _wrap_row(first_entry, rows):
+            if first_entry and (not hasattr(first_entry, '__iter__') or isinstance(first_entry, six.string_types)):
+                soft_assert_type_text(first_entry)
                 # `rows` is actually just a single row, so wrap it
                 rows = [rows]
-            ret.append(
-                (name, [cls(row) for row in rows])
-            )
-        return ret
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            if isinstance(rows, itertools.chain):
+                first_entry_rows, rows_to_wrap = itertools.tee(rows)
+                first_entry = first_entry_rows.next()
+                wrapped_rows = name, (cls(row) for row in _wrap_row(first_entry, rows_to_wrap))
+            else:
+                rows = _wrap_row(rows[0], list(rows))
+                wrapped_rows.append(
+                    (name, [cls(row) for row in rows])
```

Comment:
Doesn't this evaluate all rows in memory?

Reply:
Yes- that's ok with the chunking logic

Reply:
I do think it could be written to disk without evaluating in memory, by keeping it a generator even when writing the rows to the file.  Actually, not sure whether that's possible with our excel library, though I think it is with the csv lib.  Anyways, something to explore if this falls over with the full data set.

Reply:
Totally- that is something I briefly looked into but had trouble doing with our excel library, so I tested this implementation at icds scale on my limited release and it worked fine, so I'll opt to save that as future work if needed